### DISCUSSION TITLE
How can you set a keymap to execute lua code?

### DISCUSSION THREAD

**Discussion by MFlossmann** [nix-community/nixvim#1599] [Q&A] [answered]
I'm currently in the process of porting my [Astronvim](https://astronvim.com/) config to nixvim (I'm still not sure if I can carry through with it). Since astronvim has a lot of custom lua code for simple actions, like closing a buffer, I created a custom `myutils` plugin that's supposed to implement this complex logic.

But when I config the following in the `keymaps`:

```nix
keymaps = [
  {
    key = "<leader>c";
    action = "require('myutils.buffer').close()";
    options.desc = "Close buffer";
  }
]
```

Then, neovim doesn't actually call the lua function. The resulting mapping in the `init.lua` copies the string, when it actually should input the raw lua code:

``` lua
{
  ["action"] = "require('myutils.buffer').close()",
  ["key"] = "<leader>c",
  ["options"] = { ["desc"] = "Close buffer" },
},
```

I'm new to nix and nixvim (therefore the learning project) and the documentation didn't help me too much here, unfortunately. :(

Since this is some basic stuff, I'd have liked to define these mappings not where I define the utils plugin.

**Comment by MattSturgeon** [comment, marked as answer]
Use nixvim's "raw type":

```nix

```nix
action.__raw = "require('myutils.buffer').close";
```

Or

```nix
action.__raw = "function() require('myutils.buffer').close() end";
```

You can use `helpers.mkRaw` too (which does the same as above)

**Reply by MFlossmann** [reply]
Thanks! I found references to *something* called `__raw`, but couldn't find it in the documentation. I'd be happy to contribute a bit by extending the documentation e.g. to mention the answer to my question. 

Are contributions solely to the documentations welcome? Is there a specific process to follow, or should I just create a PR?

**Reply by MattSturgeon** [reply]
> Are contributions solely to the documentations welcome?

Definitely!

> Is there a specific process to follow, or should I just create a PR?

Just a PR, yes. Anything we're particular about can be handled in PR reviews 👍

**Reply by MattSturgeon** [reply]
We do have some mention of the raw type throughout the docs, particularly the user-guides section. But I agree it can definitely be made clearer and more obvious.

**Comment by Denperidge** [comment]
Updated answer for the current version, 26.05!
```nix
action = lib.generators.mkLuaInline ''function () ... end'';
```
